README.md
                                                                               
    888888b.                      88888888888                                  
    888  "88b                         888                                      
    888  .88P                         888                                      
    8888888K.   8888b.  88888b.d88b.  888  888  888 88888b.   .d88b.  888d888  
    888  "Y88b     "88b 888 "888 "88b 888  888  888 888 "88b d8P  Y8b 888P"    
    888    888 .d888888 888  888  888 888  888  888 888  888 88888888 888      
    888   d88P 888  888 888  888  888 888  Y88b 888 888 d88P Y8b.     888      
    8888888P"  "Y888888 888  888  888 888   "Y88888 88888P"   "Y8888  888      
                                                888 888                        
                                           Y8b d88P 888                        
                                            "Y88P"  888                         
                                                                             

# Overview

Ability to work out the orientation and insert size of a paired read data file
Can estimate relative orientation and gap between pairs of contigs in the bam file (Useful for scaffolding) 

# Installation

use pip:

```sh
$ pip install BamTyper
```
# Usage
##As a library:
###Get the type of the reads:
    #!/usr/bin/env python
    from bamtyper.utilities import BamParser as BTBP
    BP = BTBP()
    bam_types = BP.getTypes(bamFiles)

    Where:
        bamFiles - a list of BAM filenames
    
    Returns:
        bam_types - a dict containing information about the insert size and 
                    relative orientation of reads in the bam file
                      { bam1 : (type, ins, stdev), ... }
                    Where:
                        type - orientation type:
                            0 : OUT  : <--- --->
                            1 : SAME : ---> --->
                            2 : IN   : ---> <---
                        ins - estimated insert size (of original DNA fragment)
                        stdev - standard deviation of insert size      

###Get linking pairs:
    #!/usr/bin/env python
    from bamtyper.utilities import BamParser as BTBP
    BP = BTBP()
    (links, ref_lengths, coverages) = BP.getLinks(bamFiles, doCoverage=True)
    
    Where:
        bamFiles - a list of BAM filenames
    
    Returns:
        links - a dict containing information about links between two contigs
                  {c1: (c2, num_links, link_type, gap), ... }
                Where:
                    num_links - Number of paired reads confirming the link
                    link_type - Relative orientation of the two contigs (Start and End)
                      SS, SE, ES, EE or ERROR
                    gap - Estimated gap between te two contigs
                    
        coverages - a dict containing the FRAGMENT coverage of each contig n the bam file(s)
                  {c1 : (cov1, cov2, ..., covN), ... }
                  
        ref_lengths - a dict containing the lengths of all contigs
                  {c1 : len, ... }
    
    Notes:
        bamtyper will automatically work out the orientation and insert size of the 
        reads in each bam file and base it's estimations of link_type and gap on this 

##On the command line:

    bamtyper type - Parse BAM files and determine reads type
    
        $ bamtyper type bamfile.bam
        
        Determining OT for BAM 'bamfile'
        Orientation: IN Insert: 301, Stdev: 29
 
    bamtyper links - Parse BAM files and get linking reads

    Usage 1:

        $ bamtyper links bamfile.bam 

      1.
        contig2 , [ contig1 , 39 , SE , 69 ]
        contig1 , [ contig2 , 39 , ES , 69 ]
        
        implies a layout which looks like:
        
        ---1--> 69bp ---2-->

      2.
        contig3 , [ contig2 , 3 , SS , 58 ]
        contig2 , [ contig3 , 3 , SS , 58 ] , [ contig1 , 4 , EE , 45 ]
        contig1 , [ contig2 , 4 , EE , 45 ]

        implies a layout which looks like:
        
        ---1--> 45bp <--2--- 58bp ---3-->

    Usage 2: report FRAGMENT coverage too!

        $ bamtyper links bamfile1.bam bamfile2.bam -c

        contig3 , [ contig2 , 3 , SS , 58 ]
        contig2 , [ contig3 , 3 , SS , 58 ] , [ contig1 , 4 , EE , 45 ]
        contig1 , [ contig2 , 4 , EE , 45 ]
        contig3 0.6206 0.5234
        contig2 0.6558 0.0123
        contig1 0.6523 0.5634

        Where:
            contig3 0.6206 0.5234
            
            Reports 0.6206 fragments per base in bamfile1 and 0.5234 in bamfile2
            If these were 100bp reads then this would imply
            coverages of 62x and 52x respectively
            
# Administration

Project home page, info on the source tree, documentation, issues and how to contribute, see http://github.com/minillinim/BamTyper

This software is currently unpublished.

Copyright © 2012 Michael Imelfort. See LICENSE.txt for further details.



CHANGES.txt
v0.0.1, 17/10/2012 -- First build
v0.1.0, 18/10/2012 -- Working!
v0.1.1, 19/10/2012 -- Accurately estimte gaps. Doc fixups
v0.1.2, 19/10/2012 -- Small bug fix with printing bam names
v0.1.3, 19/10/2012 -- Copes with slightly overlapping contigs when looking for links
v0.1.4, 28/10/2012 -- Changed fetch behaviour to aavoid pysam issues
v0.1.5, 30/10/2012 -- Changed fetch behaviour to aavoid pysam issues again!
v0.1.6, 30/10/2012 -- allows for unpaired data
v0.2.0, 8/11/2012 -- Can extract reads or headers


bamtyper/bamtyper.py
#!/usr/bin/env python
###############################################################################
#                                                                             #
#    bamtyper.py                                                              #
#                                                                             #
#    Wraps coarse workflows                                                   #
#                                                                             #
#    Copyright (C) Michael Imelfort                                           #
#                                                                             #
###############################################################################
#                                                                             #
#  888888b.                      88888888888                                  #
#  888  "88b                         888                                      #
#  888  .88P                         888                                      #
#  8888888K.   8888b.  88888b.d88b.  888  888  888 88888b.   .d88b.  888d888  #
#  888  "Y88b     "88b 888 "888 "88b 888  888  888 888 "88b d8P  Y8b 888P"    #
#  888    888 .d888888 888  888  888 888  888  888 888  888 88888888 888      #
#  888   d88P 888  888 888  888  888 888  Y88b 888 888 d88P Y8b.     888      #
#  8888888P"  "Y888888 888  888  888 888   "Y88888 88888P"   "Y8888  888      #
#                                              888 888                        #
#                                         Y8b d88P 888                        #
#                                         "Y88P"  888                         #
#                                                                             #
###############################################################################
#                                                                             #
#    This program is free software: you can redistribute it and/or modify     #
#    it under the terms of the GNU General Public License as published by     #
#    the Free Software Foundation, either version 3 of the License, or        #
#    (at your option) any later version.                                      #
#                                                                             #
#    This program is distributed in the hope that it will be useful,          #
#    but WITHOUT ANY WARRANTY; without even the implied warranty of           #
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the            #
#    GNU General Public License for more details.                             #
#                                                                             #
#    You should have received a copy of the GNU General Public License        #
#    along with this program. If not, see <http://www.gnu.org/licenses/>.     #
#                                                                             #
###############################################################################

__author__ = "Michael Imelfort"
__copyright__ = "Copyright 2012"
__credits__ = ["Michael Imelfort"]
__license__ = "GPL3"
__version__ = "0.2.0"
__maintainer__ = "Michael Imelfort"
__email__ = "mike@mikeimelfort.com"
__status__ = "Development"

###############################################################################

import argparse
import sys
import os

# BamTyper imports
import utilities
 
###############################################################################
###############################################################################
###############################################################################
###############################################################################

class BamTyperOptionsParser():
    def __init__(self): pass
    
    def parseOptions(self, options ):

        if(options.subparser_name == 'type'):
            # Work out the read orientation type and return
            BamParser = utilities.BamParser()
            BamParser.getTypes(options.bamfiles, verbose=True)

        elif(options.subparser_name == 'extract'):
            # extract reads from the BAM files
            BamParser = utilities.BamParser()
            
            # there is only one bin in this case
            # make the list appropriate to this usage
            targets={}
            try:
                with open(options.list, "r") as lf:
                    for line in lf:
                        targets[line.rstrip()] = -1 # bid of -1 indicates single bin
            except:
                print "Could not parse list file:",options.list,exc_info()[0]
                raise
            BamParser.extractReads(options.bamfiles,
                                   options.prefix,        
                                   targets,
                                   combineBams=options.no_separate_bams,
                                   pairsOnly=options.pairs_only,
                                   combine=options.combine_reads,
                                   shuffle=options.shuffle,
                                   largeFiles=options.no_gzip,
                                   headersOnly=options.headers_only,
                                   dontTrustSamFlags=options.dont_trust_sam_flags,
                                   folder=options.out_folder
                                   )
        elif(options.subparser_name == 'links'):
            # Calculate linking reads and possibly coverages
            BamParser = utilities.BamParser()
            if options.coverage:
                (filtered_links, ref_lengths, total_coverages) = BamParser.getLinks(options.bamfiles,
                                                                                    full=options.verbose,
                                                                                    doCoverage=True,
                                                                                    minJoin=options.min)
                for cid in filtered_links:
                    print cid, 
                    for fl in filtered_links[cid]: 
                        print ", [",fl[0],",",fl[1],",",BamParser.LT2Str(fl[2], terse=True),", %0.0f ]" % fl[3],
                    print 
                for cid in ref_lengths:
                    line_vals = [cid]
                    for i in range(len(options.bamfiles)):
                        if cid in total_coverages[i]:
                            line_vals.append("%0.4f" %total_coverages[i][cid])
                        else:
                            line_vals.append("0.0")
                    print "\t".join(line_vals)
            else:
                filtered_links = BamParser.getLinks(options.bamfiles,
                                                    full=options.verbose,
                                                    doCoverage=False,
                                                    minJoin=options.min)
                for cid in filtered_links:
                    print cid, 
                    for fl in filtered_links[cid]: 
                        print ", [",fl[0],",",fl[1],",",BamParser.LT2Str(fl[2], terse=True),", %0.0f ]" % fl[3],
                    print 
        return 0

###############################################################################
###############################################################################
###############################################################################
###############################################################################
